cosmwasm/contracts/ibc-translator/src/reply.rs

```diff
@@ -0,0 +1,197 @@
+            let contract_decoded = String::from_utf8(contract.to_vec())
+                .context(format!("failed to convert {contract} to utf8 string"))?;
+            convert_cw20_to_bank_and_send(
+                deps,
+                env,
+                contract_decoded,
+                res_data.amount.into(),
+                contract_addr,
+                chain,
+                Some(payload),
+            )
+        }
+    }
+}
+pub fn convert_cw20_to_bank_and_send(
+    deps: DepsMut,
+    env: Env,
+    recipient: String,
+    amount: u128,
+    contract_addr: String,
+    chain_id: u16,
+    payload: Option<Binary>,
+) -> Result<Response<TokenFactoryMsg>, anyhow::Error> {
+    // check the recipient and contract addresses are valid
+    // recipient will have a different bech32 prefix and fail
+    //deps.api
+    //    .addr_validate(&recipient)
+    //    .context(format!("invalid recipient address {}", recipient))?;
+
+    deps.api
+        .addr_validate(&contract_addr)
+        .context(format!("invalid contract address {contract_addr}"))?;
+
+    // convert contract address into base64
+    let subdenom = contract_addr_to_base58(deps.as_ref(), contract_addr.clone())?;
+    // format the token factory denom
+    let tokenfactory_denom = "factory/".to_string()
+        + env.contract.address.to_string().as_ref()
+        + "/"
+        + subdenom.as_ref();
+
+    let mut response: Response<TokenFactoryMsg> = Response::new();
+
+    // check contract storage see if we've created a denom for this cw20 token yet
+    // if we haven't created the denom, then create the denom
+    // info.sender contains the cw20 contract address
+    if !CW_DENOMS.has(deps.storage, contract_addr.clone()) {
+        // call into token factory to create the denom
+        let create_denom = SubMsg::reply_on_success(
+            TokenMsg::CreateDenom {
+                subdenom,
+                metadata: None,
+            },
+            CREATE_DENOM_REPLY_ID,
+        );
+        response = response.add_submessage(create_denom);
+
+        // add the contract_addr => tokenfactory denom to storage
+        CW_DENOMS
+            .save(deps.storage, contract_addr, &tokenfactory_denom)
+            .context("failed to save contract_addr => tokenfactory denom to storage")?;
+    }
+
+    // add calls to mint and send bank tokens
+    response = response.add_message(TokenMsg::MintTokens {
+        denom: tokenfactory_denom.clone(),
+        amount,
+        mint_to_address: env.contract.address.to_string(),
+    });
+
+    let channel = CHAIN_TO_CHANNEL_MAP
+        .load(deps.storage, chain_id)
+        .context("chain id does not have an allowed channel")?;
+
+    let payload_decoded = match payload {
+        Some(payload) => String::from_utf8(payload.to_vec())
+            .context(format!("failed to convert {payload} to utf8 string"))?,
+        None => "".to_string(),
+    };
+
+    // Create MsgTransfer protobuf message for Stargate
```

Comment:
Let's link to the ICS20 transfer payload spec here so it's clear what protobuf object we are serializing to.